src/doorman/TelnetProtocol.cpp

```diff
@@ -0,0 +1,171 @@
+            return gsl::span<const byte>(buffer_).first(starting_point);
+        }
+        // Remove the command.
+        buffer_.erase(std::next(buffer_.begin(), starting_point),
+                      std::next(buffer_.begin(), starting_point + num_consumed));
+    }
+    return buffer_;
+}
+
+void TelnetProtocol::add_data(gsl::span<const byte> data) {
+    buffer_.insert(buffer_.end(), data.begin(), data.end());
+
+    auto span_to_scan = interpret_iacs();
+    if (span_to_scan.empty())
+        return;
+
+    auto as_string_view = std::string_view(reinterpret_cast<const char *>(span_to_scan.data()), span_to_scan.size());
+    auto remaining_data = find_whole_lines(as_string_view);
+
+    const auto num_consumed = as_string_view.size() - remaining_data.size();
+    buffer_.erase(buffer_.begin(), std::next(buffer_.begin(), num_consumed));
+}
+
+std::string_view TelnetProtocol::find_whole_lines(std::string_view data) {
+    while (!data.empty()) {
+        // Handle a \n\r or \r\n sequence; even if it crossed a buffer boundary.
+        if ((previous_separator_ == '\n' && data.front() == '\r')
+            || (previous_separator_ == '\r' && data.front() == '\n')) {
+            data.remove_prefix(1);
+            previous_separator_ = 0;
+            continue;
+        }
+
+        auto cr_or_lf = data.find_first_of("\n\r");
+        if (cr_or_lf == std::string_view::npos)
+            break;
+
+        previous_separator_ = data[cr_or_lf];
+        handler_.on_line(data.substr(0, cr_or_lf));
+        data.remove_prefix(cr_or_lf + 1);
+    }
+    return data;
+}
+
+void TelnetProtocol::set_echo(bool echo) {
+    send_com(echo ? WONT : WILL, TELOPT_ECHO);
+    echoing_ = echo;
+}
+
+void TelnetProtocol::send_telopts() {
+    send_com(DO, TELOPT_TTYPE);
+    send_com(DO, TELOPT_NAWS);
+    send_com(WONT, TELOPT_ECHO);
+}
+
+static bool is_two_byte_command(byte b) {
+    return b == NOP || b == DM || b == BREAK || b == IP || b == ABORT || b == AYT || b == EC || b == EL || b == GA
+           || b == IAC || b < SE;
+}
+
+size_t TelnetProtocol::on_subcommand(gsl::span<const byte> command_sequence) {
+    auto option_code = command_sequence[2];
+    auto body = command_sequence.subspan(3);
+    const std::array<byte, 2> iac_se = {IAC, SE};
+    auto iac_se_it = std::search(body.begin(), body.end(), iac_se.begin(), iac_se.end());
+    if (iac_se_it == body.end())
+        return 0;
+    body = body.subspan(0, std::distance(body.begin(), iac_se_it));
+    switch (option_code) {
+    case TELOPT_TTYPE:
+        if (body.size() > 1) {
+            std::string term(reinterpret_cast<const char *>(body.data() + 1), body.size() - 1);
+            ansi_ = ::supports_ansi(term.c_str());
+            handler_.on_terminal_type(term, ansi_);
+        }
+        break;
+
+    case TELOPT_NAWS:
+        if (body.size() >= 4) {
+            width_ = (body[0] << 8u) | body[1];
```

Comment:
Presumably body is a byte*. Will shifting a byte left by 8 preserve the contents because it's already implicitly turned into whatever larger int type width_ is, or do we run the risk of losing the upper 8 bits here? (same for height).

Reply:
`body[0]` is a byte (it comes from the `gsl::span<const byte>` above). Good point, my linter is complaining that this isn't obviously OK, so will change it. It _is_ ok (and even has tests!), so have changed it: thanks for being my conscience here :)